mlrun/utils.py

```diff
@@ -430,3 +431,43 @@ def fill_function_hash(function_dict, tag=''):
     function_dict['metadata']['hash'] = hashkey
     function_dict['status'] = status
     return hashkey
+
+
+def retry_until_successful(interval: int, timeout: int, logger, verbose: bool, function, *args, **kwargs):
+    """
+    Runs function with given *args and **kwargs using deferToThread.
+    Tries to run it until success or timeout reached (timeout is optional)
+    :param interval: int/float that will be used as interval
+    :param timeout: pass None if timeout is not wanted, number of seconds if it is
+    :param logger: a logger so we can log the failures
+    :param verbose: whether to log the failure on each retry
+    :param function: function to run
+    :param args: functions args
+    :param kwargs: functions kwargs
+    :return: function result
+    """
+    start_time = time.time()
+    last_exception = None
+
+    # If deadline was not provided or deadline not reached
+    while timeout is None or time.time() < start_time + timeout:
+        try:
+            result = function(*args, **kwargs)
+            return result
+
+        except Exception as exc:
+            last_exception = exc
+
+            if logger is not None and verbose:
+                logger.debug('operation not yet successful. exc: {}'.format(exc))
+
+            # If next interval is within allowed time period - wait on interval, abort otherwise
+            if timeout is None or time.time() + interval < start_time + timeout:
+                time.sleep(interval)
+            else:
+                break
+
+    if logger is not None:
+        logger.warning('operation did not complete on time. last exception: {}'.format(last_exception))
```

Comment:
Sometimes exceptions may be generic and if there are several `retry_until_successful` running in parallel it's hard to define which log belongs to which context. Try to see if the function name or something like this can be logged.
Also I think will be helpful to log the timeout